screen: Add commands to manage TV screen

camh- opened the pull request:
Add the `run` and `list` commands to manage a TV screen connected to the
X Windows server.

When `offscreen` is run with the `run` command (the default command, so
the `run` verb is not necessary), it will watch the X server for changes
to the screen saver state and propagate that state to the connected Sony
TV with a set of rules such that it works for multiple machines
connected to the TV on different inputs, each running `offscreen`.

The `list` command lists the manufacturer ID and product code of all
screens connected to the X server so that users can discover these
values for running offscreen against their Sony TVs if they have a
different model to the supported default.

juliaogris commented on cmds.go:
```diff
@@ -1,3 +1,4 @@
+//nolint:goerr113 // dynamic errors in main is OK
```
dynamic errors in main _are_ OK ?

camh- replied:
thanks. will fix

juliaogris commented on cmds.go:
```diff
@@ -16,6 +18,19 @@ import (
 var ErrUsage = errors.New("usage error")
 
+// screenFlags is a kong CLI struct to be embedded in command structs that
+// use a [Screen] struct for communicating with an X11 server. It has an
```
neat! does that automatically link to screen?!

camh- replied:
It is meant to according to https://tip.golang.org/doc/comment

juliaogris commented on cmds.go:
```diff
@@ -49,6 +72,75 @@ type SonyCmdToggle struct {
+func (cmd *RunCmd) loop(ch <-chan bool) error {
+	defer cmd.screen.Close()
+	c := NewRESTClient(cmd.Hostname, cmd.PSK)
+	ourLabel, err := hostnameLabel()
```
`ourLabel` as opposed to `theirLabel`? as in current machine label?

camh- replied:
Yes, the label of the current machine. As opposed to the labels of the other inputs the TV has

juliaogris commented on cmds.go:
```diff
@@ -49,6 +72,75 @@ type SonyCmdToggle struct {
+	}
+	cmd.Input = ourInput
+
+	for ssOn := range ch {
```
I wonder what `ss` stands for.

juliaogris replied:
ah, screensaver on.

camh- replied:
yep

juliaogris commented on cmds.go:
```diff
@@ -49,6 +72,75 @@ type SonyCmdToggle struct {
+		if err != nil {
+			return fmt.Errorf("could not get selected input: %w", err)
+		}
+		if status == "standby" && !ssOn && input != cmd.Input {
```
i'm a bit lost here. I select a different Input (computer) via cmd.Input. why does it matter what the current screensaver status is?

juliaogris replied:
i think it would be helpful to have an overview before the loop starts of the different things that can happen here.

camh- replied:
The if statements reads: If the TV was off and the screen saver has turned off and the currently selected input on the tv is not our input, the select the input that we are connected to. We previously turned on the screen so we don't need to do that. We need to do it this way because we cannot get the currently selected input while the TV is off - I thought I had a comment for that - it seems to have gotten lost with some rewriting. I'll add something back.

camh- replied:
I have updated this function, adding a bunch of comments. Let me know how it reads now. Happy to revise some more if needed.

juliaogris commented on cmds.go:
```diff
@@ -49,6 +72,75 @@ type SonyCmdToggle struct {
+	if !ok {
+		return fmt.Errorf("tv set does not have labelled input: %s", ourLabel)
+	}
+	cmd.Input = ourInput
```
wait, why do you override this? I thought this comes from the flag, what's the point of having it?

camh- replied:
yeah, thats dumb. I do rework this in the next commit and it all is handled differently. I can rework this.

camh- replied:
This ended up disappearing due to reordering of commits.

juliaogris commented on cmds.go:
```diff
@@ -49,6 +72,75 @@ type SonyCmdToggle struct {
+	ch := make(chan bool)
+	g := errgroup.Group{}
+	g.Go(func() error { return cmd.screen.Watch(ch) })
+	g.Go(func() error { return cmd.loop(ch) })
```
I don't quite understand the need for channels here.

camh- replied:
no need, just the design. it was more complex with multiple event loops as I had the screen saver events and the randr events (monitor presence) separated. but I merged them. Perhaps you are right and an interface would make this simpler. I'll try that out

juliaogris reviewed (commented):
I don't understand the wording of the second commit:

> It will turn off the TV if
the screen saver activates and the TV's current input matches that of
the host, and only when the TV is connected.

ok, this will take a bit for me to truly wrap my head around.

camh- commented:
> ok, this will take a bit for me to truly wrap my head around.

Yep. It is hard to describe sometime - the number of elements in the condition just pushes past the threshold of easily understood. Perhaps I can reword. I'll try putting something in the readme.

camh- commented:
@juliaogris I have addressed your comments. I did not update the README as I said I would, but instead have added more comments to the run function. Let me know if that makes it readable now. I still plan to update the README, but I also want to incorporate stuff around integrating offscreen with your desktop using a screen locker, etc. I need to do a little more research before I can do that.

juliaogris reviewed (approved):
🍀 - lgtm.
thanks for helping me to get more familiar with cam style range functions and atomic.Bool
